File: Project4/Ex7.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>

using namespace std;

// Cấu trúc của một nút trong cây
struct Node {
    char data;
    vector<Node*> children;
};

// Hàm tạo nút mới để code gọn hơn
Node* newNode(char data) {
    Node* node = new Node();
    node->data = data;
    return node;
}

// (a) Duyệt Tiền thứ tự (Preorder Traversal)
void preorder(Node* root) {
    if (root == nullptr) return;
    cout << root->data << " "; // 1. Thăm gốc
    for (Node* child : root->children) {
        preorder(child); // 2. Duyệt các con
    }
}

// (b) Duyệt Hậu thứ tự (Postorder Traversal)
void postorder(Node* root) {
    if (root == nullptr) return;
    for (Node* child : root->children) {
        postorder(child); // 1. Duyệt các con
    }
    cout << root->data << " "; // 2. Thăm gốc
}

// (c) Duyệt từ trên xuống (Top-down / Level-order)
void top_down(Node* root) {
    if (root == nullptr) return;
    queue<Node*> q;
    q.push(root);

    while (!q.empty()) {
        Node* current = q.front();
        q.pop();
        cout << current->data << " ";

        for (Node* child : current->children) {
            q.push(child);
        }
    }
}

// (d) Duyệt từ dưới lên (Bottom-up)
void bottom_up(Node* root) {
    if (root == nullptr) return;
    queue<Node*> q;
    stack<Node*> s;
    q.push(root);

    // Bước 1: Duyệt BFS và đẩy các nút vào stack
    while (!q.empty()) {
        Node* current = q.front();
        q.pop();
        s.push(current);

        // Chú ý: duyệt con từ phải sang trái để khi lấy ra từ stack sẽ đúng thứ tự
        for (auto it = current->children.rbegin(); it != current->children.rend(); ++it) {
             q.push(*it);
        }
    }

    // Bước 2: Lấy các nút ra khỏi stack và in
    while (!s.empty()) {
        cout << s.top()->data << " ";
        s.pop();
    }
}


int main() {
    // Xây dựng cây ví dụ
    Node* root = newNode('F');
    root->children.push_back(newNode('B'));
    root->children.push_back(newNode('G'));
    root->children.push_back(newNode('I'));
    
    root->children[0]->children.push_back(newNode('A'));
    root->children[0]->children.push_back(newNode('D'));
    
    root->children[0]->children[1]->children.push_back(newNode('C'));
    root->children[0]->children[1]->children.push_back(newNode('E'));

    root->children[2]->children.push_back(newNode('H'));

    cout << "Duyet Tien thu tu (Preorder):  ";
    preorder(root);
    cout << endl;

    cout << "Duyet Hau thu tu (Postorder):  ";
    postorder(root);
    cout << endl;

    cout << "Duyet Tu tren xuong (Top-down): ";
    top_down(root);
    cout << endl;

    cout << "Duyet Tu duoi len (Bottom-up): ";
    bottom_up(root);
    cout << endl;
    
    // Lưu ý: Cần có hàm giải phóng bộ nhớ cho cây trong ứng dụng thực tế
    return 0;
}

File: Project4/Ex7.py
import collections

class Node:
    """Định nghĩa một nút trong cây."""
    def __init__(self, data):
        self.data = data
        self.children = []

def preorder(root):
    """(a) Duyệt Tiền thứ tự (Preorder Traversal)"""
    if root is None:
        return
    print(root.data, end=" ")
    for child in root.children:
        preorder(child)

def postorder(root):
    """(b) Duyệt Hậu thứ tự (Postorder Traversal)"""
    if root is None:
        return
    for child in root.children:
        postorder(child)
    print(root.data, end=" ")

def top_down(root):
    """(c) Duyệt từ trên xuống (Top-down / Level-order)"""
    if root is None:
        return
    queue = collections.deque([root])
    while queue:
        current = queue.popleft()
        print(current.data, end=" ")
        for child in current.children:
            queue.append(child)

def bottom_up(root):
    """(d) Duyệt từ dưới lên (Bottom-up)"""
    if root is None:
        return
    queue = collections.deque([root])
    stack = []
    
    # Bước 1: Duyệt BFS và đẩy các nút vào stack
    while queue:
        current = queue.popleft()
        stack.append(current)
        # Thêm con vào queue từ phải sang trái
        for child in reversed(current.children):
            queue.append(child)
    
    # Bước 2: Lấy các nút ra khỏi stack và in
    while stack:
        print(stack.pop().data, end=" ")

if __name__ == "__main__":
    # Xây dựng cây ví dụ
    root = Node('F')
    b_node = Node('B')
    g_node = Node('G')
    i_node = Node('I')
    root.children.extend([b_node, g_node, i_node])
    
    a_node = Node('A')
    d_node = Node('D')
    b_node.children.extend([a_node, d_node])
    
    c_node = Node('C')
    e_node = Node('E')
    d_node.children.extend([c_node, e_node])
    
    h_node = Node('H')
    i_node.children.append(h_node)

    print("Duyet Tien thu tu (Preorder):  ", end="")
    preorder(root)
    print()

    print("Duyet Hau thu tu (Postorder):  ", end="")
    postorder(root)
    print()

    print("Duyet Tu tren xuong (Top-down): ", end="")
    top_down(root)
    print()

    print("Duyet Tu duoi len (Bottom-up): ", end="")
    bottom_up(root)
    print()